# Test 1 

In [10]:
!tempo2 -f J0835-4510.par -polyco "60314 60314  60 5 12.0 GMRT 320" -tempo1 #Full day of observation

This program comes with ABSOLUTELY NO WARRANTY.
This is free software, and you are welcome to redistribute it
under conditions of GPL license.

************************************************* 
************************************************* 
[tempo2Util.C:396] Warning: [CLK3] no clock corrections available for clock UTC(gmrt) for MJD 60313.4
[tempo2Util.C:396] Warning: [CLK4] Trying assuming UTC = UTC(gmrt)
[tempo2Util.C:396] Warning: [CLK9] ... ok, using stated approximation 
[tempo2Util.C:401] Warning: [DUP1] duplicated warnings have been suppressed.
[tempo2Util.C:396] Warning: [CLK10] MJD is outside of UT1 table range  (MJD 60313.420108)
[tempo2Util.C:396] Warning: [CLK6] Proceeding assuming UTC =  UTC(gmrt)


In [11]:
!cat polyco_new.dat

0835-4510   4-Jan-24   93000.00   60313.39583333330            67.844623 -0.386 -7.672
   3816844432.039451   11.188233045354 GMRT   60    5   320.000                
  1.01988017900468109e-08 -2.87528589905144683e-01  1.33088276307748115e-06
 -4.20582342457158250e-10 -2.38130387172218593e-12
0835-4510   4-Jan-24  103000.00   60313.43749999990            67.844623 -0.388 -7.921
   3816884692.431304   11.188233045354 GMRT   60    5   320.000                
 -9.31822883834928229e-09 -2.87375459142165002e-01  1.20479101103670976e-06
 -9.69782955175301044e-10 -2.14985672748385134e-12
0835-4510   4-Jan-24  113000.00   60313.47916666660            67.844623 -0.390 -7.901
   3816924952.831872   11.188233045354 GMRT   60    5   320.000                
  9.90145693874077123e-09 -2.87243174890621233e-01  9.85469762666816238e-07
 -1.45097548050927919e-09 -1.83733835144827671e-12
0835-4510   4-Jan-24  123000.00   60313.52083333330            67.844623 -0.392 -7.669
   3816965213.239488   11.18823

In [12]:
from astropy.time import Time, TimeDelta

#Lets say you want to observe after every 10 min for an hour after the transit at GMRT 
dt_observation = 60 #min
dT_duration_of_observation = 12#hour
No_of_observation = ((int)(float(float(120.0)*dT_duration_of_observation+dt_observation-1)/float(dt_observation))); #polyco command 

observation_mjds_values = []
for i in range(int(No_of_observation)):
    observation_mjds = 60313.39583333330 +(dt_observation/1440)*i
    time_object = Time( observation_mjds, format='mjd')
    datetime_object = time_object.datetime
    #print(f"The UTC at observation is :{datetime_object}") #Date and time : UTC 
    observation_mjds_values.append(observation_mjds)

print(f"The no. of outputs in the polyco command:{No_of_observation}")
print(f"MJD values at which phase can be measured is : {observation_mjds_values}")

The no. of outputs in the polyco command:24
MJD values at which phase can be measured is : [60313.3958333333, 60313.43749999996, 60313.479166666635, 60313.5208333333, 60313.56249999996, 60313.604166666635, 60313.6458333333, 60313.68749999996, 60313.729166666635, 60313.7708333333, 60313.81249999996, 60313.854166666635, 60313.8958333333, 60313.93749999996, 60313.979166666635, 60314.0208333333, 60314.06249999996, 60314.104166666635, 60314.1458333333, 60314.18749999996, 60314.229166666635, 60314.2708333333, 60314.31249999996, 60314.354166666635]


In [18]:
import datetime
import math
for mjd_obs in observation_mjds_values:
    # Given values
    ra_str = "08h35m20.5554429s" #RA of Pulsar 
    afmjd = mjd_obs  # mjd1_gmrt_round_new    #60268 #60297  # Example value for afmjd(MJD_start)
    maxha = 12 #  maxha 
    #tspan = 60  #  tspan (min)

    # Observatory coordinates  1656342.30    5797947.77 
    obs_x =  1656342.30  # Replace with the actual x-coordinate
    obs_y = 5797947.77 # Replace with the actual y-coordinate

    # Convert transit LST to datetime
    #transit_lst_time = datetime.datetime.strptime(transit_lst_str, "%H:%M:%S").time() 60312.97916666660

    # Convert RA to hours
    ra_hours = (8 + (35/60) + (20.5554429/3600))

    # Conversion of UT into local sidereal time
    solsid = 1.002737909

    #  LST transit in hours (taken from the NCRA transit time calculator)
    #hlst_gmrt = transit_lst_time.hour + transit_lst_time.minute/60 + transit_lst_time.second/3600 

    # Calculate observatory longitude
    alng = math.atan2(-obs_y, obs_x)

    # Conversion of UT into local sidereal time
    solsid = 1.002737909
    hlst = 24.0 * ((solsid * afmjd + 0.154374 - alng / (2 * math.pi)) % 1.0)

    
    rax = int(ra_hours) + int((ra_hours - int(ra_hours)) * 60.0 + 2) / 60.0

    wait_gmrt = (rax - hlst ) / solsid #GMRT # Calculate wait : Local hour angle

    if wait_gmrt < -maxha:
        wait_gmrt += 24.0/solsid

    print(f"TMID:{mjd_obs} HA:{wait_gmrt}")

TMID:60313.3958333333 HA:11.261610781450134
TMID:60313.43749999996 HA:10.261610781476701
TMID:60313.479166666635 HA:9.261610781329125
TMID:60313.5208333333 HA:8.261610781355692
TMID:60313.56249999996 HA:7.261610781382262
TMID:60313.604166666635 HA:6.261610781234684
TMID:60313.6458333333 HA:5.261610781435399
TMID:60313.68749999996 HA:4.261610781461967
TMID:60313.729166666635 HA:3.2616107813143898
TMID:60313.7708333333 HA:2.2616107813409587
TMID:60313.81249999996 HA:1.2616107813675272
TMID:60313.854166666635 HA:0.26161078139409577
TMID:60313.8958333333 HA:-0.7383892185793356
TMID:60313.93749999996 HA:-1.738389218552767
TMID:60313.979166666635 HA:-2.7383892187003447
TMID:60314.0208333333 HA:-3.738389218673776
TMID:60314.06249999996 HA:-4.738389218647208
TMID:60314.104166666635 HA:-5.738389218620639
TMID:60314.1458333333 HA:-6.73838921859407
TMID:60314.18749999996 HA:-7.738389218567502
TMID:60314.229166666635 HA:-8.73838921871508
TMID:60314.2708333333 HA:-9.73838921868851
TMID:60314.312499

# Test 2 

In [19]:
!tempo2 -f J0835-4510.par -polyco "60314 60314  60 5 6.0 GMRT 320" -tempo1

1052.19s - pydevd: Sending message related to process being replaced timed-out after 5 seconds
This program comes with ABSOLUTELY NO WARRANTY.
This is free software, and you are welcome to redistribute it
under conditions of GPL license.

************************************************* 
************************************************* 
[tempo2Util.C:396] Warning: [CLK3] no clock corrections available for clock UTC(gmrt) for MJD 60313.7
[tempo2Util.C:396] Warning: [CLK4] Trying assuming UTC = UTC(gmrt)
[tempo2Util.C:396] Warning: [CLK9] ... ok, using stated approximation 
[tempo2Util.C:401] Warning: [DUP1] duplicated warnings have been suppressed.
[tempo2Util.C:396] Warning: [CLK10] MJD is outside of UT1 table range  (MJD 60313.670108)
[tempo2Util.C:396] Warning: [CLK6] Proceeding assuming UTC =  UTC(gmrt)


In [20]:
!cat polyco_new.dat 

1071.22s - pydevd: Sending message related to process being replaced timed-out after 5 seconds
0835-4510   4-Jan-24  153000.00   60313.64583333330            67.844623 -0.392 -7.669
   3817085994.481322   11.188233045354 GMRT   60    5   320.000                
  1.25343062178364624e-08 -2.87094357946420742e-01 -4.52865874276856444e-07
 -2.16485429805186532e-09  4.49497515705278760e-13
0835-4510   4-Jan-24  163000.00   60313.68749999990            67.844623 -0.391 -7.891
   3817126254.892469   11.188233045354 GMRT   60    5   320.000                
 -1.01227387058442214e-08 -2.87171585926344453e-01 -8.28696844553204301e-07
 -1.98037027587185492e-09  1.09047619611646118e-12
0835-4510   4-Jan-24  173000.00   60313.72916666660            67.844623 -0.390 -7.894
   3817166515.297773   11.188233045354 GMRT   60    5   320.000                
 -9.58283135919137175e-10 -2.87291395449289610e-01 -1.15844715150209195e-06
 -1.65818916070120061e-09  1.56977376312766977e-12
0835-4510   4-Jan-24  1

In [21]:
from astropy.time import Time, TimeDelta

#Lets say you want to observe after every 10 min for an hour after the transit at GMRT 
dt_observation = 60 #min
dT_duration_of_observation = 6#hour
No_of_observation = ((int)(float(float(120.0)*dT_duration_of_observation+dt_observation-1)/float(dt_observation))); #polyco command 

observation_mjds_values = []
for i in range(int(No_of_observation)):
    observation_mjds = 60313.64583333330  +(dt_observation/1440)*i
    time_object = Time( observation_mjds, format='mjd')
    datetime_object = time_object.datetime
    #print(f"The UTC at observation is :{datetime_object}") #Date and time : UTC 
    observation_mjds_values.append(observation_mjds)

print(f"The no. of outputs in the polyco command:{No_of_observation}")
print(f"MJD values at which phase can be measured is : {observation_mjds_values}")

The no. of outputs in the polyco command:12
MJD values at which phase can be measured is : [60313.6458333333, 60313.68749999996, 60313.729166666635, 60313.7708333333, 60313.81249999996, 60313.854166666635, 60313.8958333333, 60313.93749999996, 60313.979166666635, 60314.0208333333, 60314.06249999996, 60314.104166666635]


In [22]:
import datetime
import math
for mjd_obs in observation_mjds_values:
    # Given values
    ra_str = "08h35m20.5554429s" #RA of Pulsar 
    afmjd = mjd_obs  # mjd1_gmrt_round_new    #60268 #60297  # Example value for afmjd(MJD_start)
    maxha = 12 #  maxha 
    #tspan = 60  #  tspan (min)

    # Observatory coordinates  1656342.30    5797947.77 
    obs_x =  1656342.30  # Replace with the actual x-coordinate
    obs_y = 5797947.77 # Replace with the actual y-coordinate

    # Convert transit LST to datetime
    #transit_lst_time = datetime.datetime.strptime(transit_lst_str, "%H:%M:%S").time() 60312.97916666660

    # Convert RA to hours
    ra_hours = (8 + (35/60) + (20.5554429/3600))

    # Conversion of UT into local sidereal time
    solsid = 1.002737909

    #  LST transit in hours (taken from the NCRA transit time calculator)
    #hlst_gmrt = transit_lst_time.hour + transit_lst_time.minute/60 + transit_lst_time.second/3600 

    # Calculate observatory longitude
    alng = math.atan2(-obs_y, obs_x)

    # Conversion of UT into local sidereal time
    solsid = 1.002737909
    hlst = 24.0 * ((solsid * afmjd + 0.154374 - alng / (2 * math.pi)) % 1.0)

    
    rax = int(ra_hours) + int((ra_hours - int(ra_hours)) * 60.0 + 2) / 60.0

    wait_gmrt = (rax - hlst ) / solsid #GMRT # Calculate wait : Local hour angle

    if wait_gmrt < -maxha:
        wait_gmrt += 24.0/solsid

    print(f"TMID:{mjd_obs} HA:{wait_gmrt}")

TMID:60313.6458333333 HA:5.261610781435399
TMID:60313.68749999996 HA:4.261610781461967
TMID:60313.729166666635 HA:3.2616107813143898
TMID:60313.7708333333 HA:2.2616107813409587
TMID:60313.81249999996 HA:1.2616107813675272
TMID:60313.854166666635 HA:0.26161078139409577
TMID:60313.8958333333 HA:-0.7383892185793356
TMID:60313.93749999996 HA:-1.738389218552767
TMID:60313.979166666635 HA:-2.7383892187003447
TMID:60314.0208333333 HA:-3.738389218673776
TMID:60314.06249999996 HA:-4.738389218647208
TMID:60314.104166666635 HA:-5.738389218620639


# Test 3 

In [23]:
!tempo2 -f J0835-4510.par -polyco "60314 60315  60 5 12.0 GMRT 320" -tempo1

This program comes with ABSOLUTELY NO WARRANTY.
This is free software, and you are welcome to redistribute it
under conditions of GPL license.

************************************************* 
************************************************* 
[tempo2Util.C:396] Warning: [CLK3] no clock corrections available for clock UTC(gmrt) for MJD 60313.4
[tempo2Util.C:396] Warning: [CLK4] Trying assuming UTC = UTC(gmrt)
[tempo2Util.C:396] Warning: [CLK9] ... ok, using stated approximation 
[tempo2Util.C:401] Warning: [DUP1] duplicated warnings have been suppressed.
[tempo2Util.C:396] Warning: [CLK10] MJD is outside of UT1 table range  (MJD 60313.420108)
[tempo2Util.C:396] Warning: [CLK6] Proceeding assuming UTC =  UTC(gmrt)


In [24]:
!cat polyco_new.dat

0835-4510   4-Jan-24   93000.00   60313.39583333330            67.844623 -0.386 -7.672
   3816844432.039451   11.188233045354 GMRT   60    5   320.000                
  1.01988017900468109e-08 -2.87528589905144683e-01  1.33088276307748115e-06
 -4.20582342457158250e-10 -2.38130387172218593e-12
0835-4510   4-Jan-24  103000.00   60313.43749999990            67.844623 -0.388 -7.921
   3816884692.431304   11.188233045354 GMRT   60    5   320.000                
 -9.31822883834928229e-09 -2.87375459142165002e-01  1.20479101103670976e-06
 -9.69782955175301044e-10 -2.14985672748385134e-12
0835-4510   4-Jan-24  113000.00   60313.47916666660            67.844623 -0.390 -7.901
   3816924952.831872   11.188233045354 GMRT   60    5   320.000                
  9.90145693874077123e-09 -2.87243174890621233e-01  9.85469762666816238e-07
 -1.45097548050927919e-09 -1.83733835144827671e-12
0835-4510   4-Jan-24  123000.00   60313.52083333330            67.844623 -0.392 -7.669
   3816965213.239488   11.18823

 Here it can be seen that tempo2 has added almost a day to understand why this was done we check the local HA at the start MJD : 60314 
 On observation: 
   1. In tempo2 the range of MJD at which the phases will be calculated is decided by a formula given in the report. 
   2. There we can see that it creates an offset and that offset decides the range in which polyco outputs are generated 

In [23]:
# Given values
import math
ra_str = "08h35m20.5554429s" #RA of Pulsar 
afmjd = 60314 # mjd1_gmrt_round_new    #60268 #60297  # Example value for afmjd(MJD_start)
maxha = 12 #  maxha 
tspan = 60  #  tspan (min)

# Observatory coordinates  1656342.30    5797947.77 
obs_x =  1656342.30  # Replace with the actual x-coordinate
obs_y = 5797947.77 # Replace with the actual y-coordinate

# Convert transit LST to datetime
#transit_lst_time = datetime.datetime.strptime(transit_lst_str, "%H:%M:%S").time() 60312.97916666660

# Convert RA to hours
ra_hours = (8 + (35/60) + (20.5554429/3600))

# Conversion of UT into local sidereal time
solsid = 1.002737909

#  LST transit in hours (taken from the NCRA transit time calculator)
#hlst_gmrt = transit_lst_time.hour + transit_lst_time.minute/60 + transit_lst_time.second/3600 

# Calculate observatory longitude
alng = math.atan2(-obs_y, obs_x)

# Conversion of UT into local sidereal time
solsid = 1.002737909
hlst = 24.0 * ((solsid * afmjd + 0.154374 - alng / (2 * math.pi)) % 1.0)


rax = int(ra_hours) + int((ra_hours - int(ra_hours)) * 60.0 + 2) / 60.0

wait_gmrt = (rax - hlst ) / solsid #GMRT # Calculate wait : Local hour angle

if wait_gmrt < -maxha:
    wait_gmrt += 24.0/solsid

wait_gmrt
offset = ((wait_gmrt - maxha) / float(24.0)) + (tspan / float(2 * 1440.0))
print(f"The offset for MJD_start {afmjd} adjusted by polyco is {offset} and HA is {wait_gmrt}")

The offset for MJD_start 60314 adjusted by polyco is -0.6140995508076519 and HA is -3.238389219383645


In [15]:
60313.38590044919 + (tspan / float(2 * 1440.0))

60313.406733782525

In [16]:
(60313.38590044919+60313.406733782525)/2 #close to tmid maybe this is why it is called tmid 

60313.396317115854

In [14]:
afmjd + offset 

60313.38590044919

So here we can observe that when start MJD of 60314 is given Polyco will create an offset of -0.614099 days which is 60313.38590044919 and is rounded off to 60313.39583333330

In [20]:
# for reference I have written a python equivalent of the roundoff which presently is not clear to me 

mjd_gmrt= 60313.38590044919#desired output
mjd_gmrt_frac = round(48.0 * (mjd_gmrt - int(mjd_gmrt))) / 48.0
ntmp1 = int(mjd_gmrt_frac * 1.0e8)
ntmp2 = int((mjd_gmrt_frac * 1.0e8 - ntmp1) * 100.0)
mjd1 = ntmp2 * 1e-10 + ntmp1 * 1.0e-8
mjd_gmrt_frac=mjd1
mjd_gmrt_transit_round = int(mjd_gmrt) + mjd_gmrt_frac#round(48.0 * (mjd1_gmrt - int(mjd1_gmrt))) / 48.0
#mjd1_gmrt_round_new = int(mjd1_gmrt) + mjd1_gmrt_frac_new
print(mjd_gmrt_transit_round) #This should match with the first MJD in polyco output 

60313.3958333333


So now the possible values which includes the values from 60313.3958333333 is added to predicted 24 hours so let us check the HA at each Polyco output but there is a problem. 

In [24]:
from astropy.time import Time, TimeDelta


dt_observation = 60 #min
dT_duration_of_observation = 12#hour
No_of_observation = 2*((int)(float(float(120.0)*dT_duration_of_observation+dt_observation-1)/float(dt_observation))); #polyco command 

observation_mjds_values = []
for i in range(int(No_of_observation)):
    observation_mjds = 60313.3958333333  +(dt_observation/1440)*i
    time_object = Time( observation_mjds, format='mjd')
    datetime_object = time_object.datetime
    #print(f"The UTC at observation is :{datetime_object}") #Date and time : UTC 
    observation_mjds_values.append(observation_mjds)

print(f"The no. of outputs in the polyco command:{No_of_observation}")
print(f"MJD values at which phase can be measured is : {observation_mjds_values}")

The no. of outputs in the polyco command:48
MJD values at which phase can be measured is : [60313.3958333333, 60313.43749999996, 60313.479166666635, 60313.5208333333, 60313.56249999996, 60313.604166666635, 60313.6458333333, 60313.68749999996, 60313.729166666635, 60313.7708333333, 60313.81249999996, 60313.854166666635, 60313.8958333333, 60313.93749999996, 60313.979166666635, 60314.0208333333, 60314.06249999996, 60314.104166666635, 60314.1458333333, 60314.18749999996, 60314.229166666635, 60314.2708333333, 60314.31249999996, 60314.354166666635, 60314.3958333333, 60314.43749999996, 60314.479166666635, 60314.5208333333, 60314.56249999996, 60314.604166666635, 60314.6458333333, 60314.68749999996, 60314.729166666635, 60314.7708333333, 60314.81249999996, 60314.854166666635, 60314.8958333333, 60314.93749999996, 60314.979166666635, 60315.0208333333, 60315.06249999996, 60315.104166666635, 60315.1458333333, 60315.18749999996, 60315.229166666635, 60315.2708333333, 60315.31249999996, 60315.3541666666

The above array clearly does not work as there is some values in the middle that is bing adjusted by polyco as they could be away from the HA range so to understand that let us calculate the HA at each tmid for that let us use pinta 

In [ ]:
from pint.models.timing_model import Component
import pint.config
import pint.logging
from pint.polycos import Polycos
import pint.toa as toa

import numpy as np 
plc = pint.polycos.tempo_polyco_table_reader(filename="polyco_new.dat") #to read the polyco file and write it as a table whic is much more flexible to use 

#print(plc) if you want to print the pint table for polycos

In [21]:
observation_mjds_values = plc['tmid'] #this gives the tmids from the polyco file 

In [22]:
import datetime
import math
for mjd_obs in observation_mjds_values:
    # Given values
    ra_str = "08h35m20.5554429s" #RA of Pulsar 
    afmjd = mjd_obs  # mjd1_gmrt_round_new    #60268 #60297  # Example value for afmjd(MJD_start)
    maxha = 12 #  maxha 
    tspan = 60  #  tspan (min)

    # Observatory coordinates  1656342.30    5797947.77 
    obs_x =  1656342.30  # Replace with the actual x-coordinate
    obs_y = 5797947.77 # Replace with the actual y-coordinate

    # Convert transit LST to datetime
    #transit_lst_time = datetime.datetime.strptime(transit_lst_str, "%H:%M:%S").time() 60312.97916666660

    # Convert RA to hours
    ra_hours = (8 + (35/60) + (20.5554429/3600))

    # Conversion of UT into local sidereal time
    solsid = 1.002737909

    #  LST transit in hours (taken from the NCRA transit time calculator)
    #hlst_gmrt = transit_lst_time.hour + transit_lst_time.minute/60 + transit_lst_time.second/3600 

    # Calculate observatory longitude
    alng = math.atan2(-obs_y, obs_x)

    # Conversion of UT into local sidereal time
    solsid = 1.002737909
    hlst = 24.0 * ((solsid * afmjd + 0.154374 - alng / (2 * math.pi)) % 1.0)

    
    rax = int(ra_hours) + int((ra_hours - int(ra_hours)) * 60.0 + 2) / 60.0

    wait_gmrt = (rax - hlst ) / solsid #GMRT # Calculate wait : Local hour angle

    if wait_gmrt < -maxha:
        wait_gmrt += 24.0/solsid

    print(f"TMID:{mjd_obs} HA:{wait_gmrt}")

TMID:60313.3958333333 HA:11.261610781381343
TMID:60313.4374999999 HA:10.261610782981263
TMID:60313.4791666666 HA:9.261610782181318
TMID:60313.5208333333 HA:8.261610781381288
TMID:60313.5624999999 HA:7.261610782981294
TMID:60313.6041666666 HA:6.261610782181349
TMID:60313.6458333333 HA:5.261610781381318
TMID:60313.6874999999 HA:4.261610782981325
TMID:60313.7291666666 HA:3.2616107821812945
TMID:60313.7708333333 HA:2.261610781381349
TMID:60313.8124999999 HA:1.2616107829812706
TMID:60313.8541666666 HA:0.261610782181325
TMID:60313.8958333333 HA:-0.7383892186187055
TMID:60313.9374999999 HA:-1.738389217018699
TMID:60313.9791666666 HA:-2.7383892178186446
TMID:60314.0208333333 HA:-3.738389218618675
TMID:60314.0624999999 HA:-4.7383892170186686
TMID:60314.1041666666 HA:-5.738389217818699
TMID:60314.1458333333 HA:-6.738389218618645
TMID:60314.1874999999 HA:-7.738389217018723
TMID:60314.2291666666 HA:-8.738389217818668
TMID:60314.2708333333 HA:-9.7383892186187
TMID:60314.3124999999 HA:-10.7383892170

After observing the above array one can clearly how polyco is adjusting to get the values within the +HA and -HA range and thus once can understand why there are sudden jumps in the output. 

The offset value that we calculted in the previous step plays a role to decide the array of mjds by Polyco now we could be careful to give a start MJD having less offset which takes us to our next test . The goal is that we require values from 60314 so values from 60313 is redundant to us which is printed because of the intitiall offset . 

# Test 4

In [1]:
#Lets us try  60314.6 

#let us check the Offset 
# Given values
import math
ra_str = "08h35m20.5554429s" #RA of Pulsar 
afmjd = 60314.6 # mjd1_gmrt_round_new    #60268 #60297  # Example value for afmjd(MJD_start)
maxha = 12 #  maxha 
tspan = 60  #  tspan (min)

# Observatory coordinates  1656342.30    5797947.77 
obs_x =  1656342.30  # Replace with the actual x-coordinate
obs_y = 5797947.77 # Replace with the actual y-coordinate

# Convert transit LST to datetime
#transit_lst_time = datetime.datetime.strptime(transit_lst_str, "%H:%M:%S").time() 60312.97916666660

# Convert RA to hours
ra_hours = (8 + (35/60) + (20.5554429/3600))

# Conversion of UT into local sidereal time
solsid = 1.002737909

#  LST transit in hours (taken from the NCRA transit time calculator)
#hlst_gmrt = transit_lst_time.hour + transit_lst_time.minute/60 + transit_lst_time.second/3600 

# Calculate observatory longitude
alng = math.atan2(-obs_y, obs_x)

# Conversion of UT into local sidereal time
solsid = 1.002737909
hlst = 24.0 * ((solsid * afmjd + 0.154374 - alng / (2 * math.pi)) % 1.0)


rax = int(ra_hours) + int((ra_hours - int(ra_hours)) * 60.0 + 2) / 60.0

wait_gmrt = (rax - hlst ) / solsid #GMRT # Calculate wait : Local hour angle

if wait_gmrt < -maxha:
    wait_gmrt += 24.0/solsid

wait_gmrt
offset = ((wait_gmrt - maxha) / float(24.0)) + (tspan / float(2 * 1440.0))
print(f"The offset for MJD_start {afmjd} adjusted by polyco is {offset} days and HA is {wait_gmrt}")

The offset for MJD_start 60314.6 adjusted by polyco is -0.21682998413115895 and HA is 6.296080380852185


In [2]:
!tempo2 -f J0835-4510.par -polyco "60314.6 60315  60 5 12.0 GMRT 320" -tempo1

This program comes with ABSOLUTELY NO WARRANTY.
This is free software, and you are welcome to redistribute it
under conditions of GPL license.

************************************************* 
************************************************* 
[tempo2Util.C:396] Warning: [CLK3] no clock corrections available for clock UTC(gmrt) for MJD 60314.4
[tempo2Util.C:396] Warning: [CLK4] Trying assuming UTC = UTC(gmrt)
[tempo2Util.C:396] Warning: [CLK9] ... ok, using stated approximation 
[tempo2Util.C:401] Warning: [DUP1] duplicated warnings have been suppressed.
[tempo2Util.C:396] Warning: [CLK10] MJD is outside of UT1 table range  (MJD 60314.399274)
[tempo2Util.C:396] Warning: [CLK6] Proceeding assuming UTC =  UTC(gmrt)


In [3]:
!cat polyco_new.dat

0835-4510   5-Jan-24   90000.00   60314.37500000000            67.844623 -0.379 -7.675
   3817790550.637362   11.188233045354 GMRT   60    5   320.000                
  1.12774094446837777e-08 -2.88053343925655347e-01  1.35128109696584956e-06
 -1.71171160854191872e-10 -2.42565297599917074e-12
0835-4510   5-Jan-24  100000.00   60314.41666666660            67.844623 -0.381 -7.928
   3817830810.997856   11.188233045354 GMRT   60    5   320.000                
 -8.38819783531023163e-09 -2.87895117010999790e-01  1.26880705298265159e-06
 -7.37808854560713355e-10 -2.25156306829195167e-12
0835-4510   5-Jan-24  110000.00   60314.45833333330            67.844623 -0.384 -7.891
   3817871071.367524   11.188233045354 GMRT   60    5   320.000                
  1.07818361936770323e-08 -2.87752747048176882e-01  1.08854891609063446e-06
 -1.25193328882664700e-09 -2.00710164459583473e-12
0835-4510   5-Jan-24  120000.00   60314.50000000000            67.844623 -0.385 -7.672
   3817911331.744978   11.18823

In [4]:
from astropy.time import Time, TimeDelta

 
dt_observation = 60 #min
dT_duration_of_observation = 12#hour
No_of_observation = ((int)(float(float(120.0)*dT_duration_of_observation+dt_observation-1)/float(dt_observation))); #polyco command 

observation_mjds_values = []
for i in range(int(No_of_observation)):
    observation_mjds = 60314.37500000000  +(dt_observation/1440)*i #use the first MJD value from polyco_new.dat 
    time_object = Time( observation_mjds, format='mjd')
    datetime_object = time_object.datetime
    #print(f"The UTC at observation is :{datetime_object}") #Date and time : UTC 
    observation_mjds_values.append(observation_mjds)

print(f"The no. of outputs in the polyco command:{No_of_observation}")
print(f"MJD values at which phase can be measured is : {observation_mjds_values}")

The no. of outputs in the polyco command:24
MJD values at which phase can be measured is : [60314.375, 60314.416666666664, 60314.458333333336, 60314.5, 60314.541666666664, 60314.583333333336, 60314.625, 60314.666666666664, 60314.708333333336, 60314.75, 60314.791666666664, 60314.833333333336, 60314.875, 60314.916666666664, 60314.958333333336, 60315.0, 60315.041666666664, 60315.083333333336, 60315.125, 60315.166666666664, 60315.208333333336, 60315.25, 60315.291666666664, 60315.333333333336]


The above array of MJDS matches exactly to what we would expect , observe the polyco mjd values. Now let us check the HA 

In [5]:
import datetime
import math
for mjd_obs in observation_mjds_values:
    # Given values
    ra_str = "08h35m20.5554429s" #RA of Pulsar 
    afmjd = mjd_obs  # mjd1_gmrt_round_new    #60268 #60297  # Example value for afmjd(MJD_start)
    maxha = 12 #  maxha 
    #tspan = 60  #  tspan (min)

    # Observatory coordinates  1656342.30    5797947.77 
    obs_x =  1656342.30  # Replace with the actual x-coordinate
    obs_y = 5797947.77 # Replace with the actual y-coordinate

    # Convert transit LST to datetime
    #transit_lst_time = datetime.datetime.strptime(transit_lst_str, "%H:%M:%S").time() 60312.97916666660

    # Convert RA to hours
    ra_hours = (8 + (35/60) + (20.5554429/3600))

    # Conversion of UT into local sidereal time
    solsid = 1.002737909

    #  LST transit in hours (taken from the NCRA transit time calculator)
    #hlst_gmrt = transit_lst_time.hour + transit_lst_time.minute/60 + transit_lst_time.second/3600 

    # Calculate observatory longitude
    alng = math.atan2(-obs_y, obs_x)

    # Conversion of UT into local sidereal time
    solsid = 1.002737909
    hlst = 24.0 * ((solsid * afmjd + 0.154374 - alng / (2 * math.pi)) % 1.0)

    
    rax = int(ra_hours) + int((ra_hours - int(ra_hours)) * 60.0 + 2) / 60.0

    wait_gmrt = (rax - hlst ) / solsid #GMRT # Calculate wait : Local hour angle

    if wait_gmrt < -maxha:
        wait_gmrt += 24.0/solsid

    print(f"TMID:{mjd_obs} HA:{wait_gmrt}")

TMID:60314.375 HA:11.696080380778373
TMID:60314.416666666664 HA:10.696080380804942
TMID:60314.458333333336 HA:9.69608038083151
TMID:60314.5 HA:8.69608038085808
TMID:60314.541666666664 HA:7.696080380884648
TMID:60314.583333333336 HA:6.69608038073707
TMID:60314.625 HA:5.696080380763639
TMID:60314.666666666664 HA:4.696080380964354
TMID:60314.708333333336 HA:3.6960803808167757
TMID:60314.75 HA:2.6960803808433442
TMID:60314.791666666664 HA:1.696080380869913
TMID:60314.833333333336 HA:0.6960803807223354
TMID:60314.875 HA:-0.303919619251096
TMID:60314.916666666664 HA:-1.3039196190503812
TMID:60314.958333333336 HA:-2.3039196191979587
TMID:60315.0 HA:-3.30391961917139
TMID:60315.041666666664 HA:-4.303919619144821
TMID:60315.083333333336 HA:-5.303919619292399
TMID:60315.125 HA:-6.303919619091684
TMID:60315.166666666664 HA:-7.303919619065116
TMID:60315.208333333336 HA:-8.303919619212694
TMID:60315.25 HA:-9.303919619186125
TMID:60315.291666666664 HA:-10.303919619159556
TMID:60315.333333333336 HA:-